Change the location of the vswhere download

adiaaida opened the pull request:
This change changes where we look to download vswhere from the github releases page, which seems to have added flakiness to builds to azure blob storage.

Storage now contains 2.5.2 (the default version), and 2.2.7, which roslyn-analyzers appears to specify in their global.json.

MattGal commented:
out of curiosity why not add retry in this PR?

adiaaida commented:
Good point. Retry capabilities have been added by using Get-File.

adiaaida commented:
@jaredpar FYI, this change should help eliminate any failures where you're seeing errors like this:

```
Downloading vswhere
The underlying connection was closed: An unexpected error occurred on a send.
```

MattGal commented:
> I don't like adding dependency on native tools. The dependency direction should be the other way around.

Can you expand on this? We bootstrap many things from our own controlled storage account.  It receives far less traffic than actual github and allows us locked in control over the version we use.

adiaaida commented:
I think what @tmat is concerned about is using native/CommonLibraries.psm1 (I might be wrong). I added that dependency because that is where we have the retry logic and I didn't want to reimplement it when it was already implemented. As this is a native tool, I felt using the logic was appropriate.

tmat commented:
> I think what @tmat is concerned about is using native/CommonLibraries.psm1

Yes, that's what I meant. A bunch of repositories use vswhere to locate VS, but almost none of them need native tools bootstrapping infrastructure. In fact, the native tools bootstrapper can't be used to install VS if it is not present on the machine.

I suggest moving general-purpose functions (that are not specific to native tools) from `common/native/CommonLibraries.psm1|sh` to `common/utils.ps1|sh` and including this file from both `tools.ps1|sh` and native scripts that need them. Similarly to `common/pipeline-logging-functions.ps1`. We might also want to consider merging `pipeline-logging-functions.ps1|sh` with `utils.ps1|sh`.

Also, I'm not sure what benefit does it have to have these wrapped in a PowerShell module, while we don't use modules for anything else. It seems to just introduce unnecessary differences between PowerShell and Bash scripts.

adiaaida commented:
I think that a restructure is out of the scope of this change. I'd like to get @chcosta's thoughts on the following, since he created the file and probably has the most context (he is oof today):

1. Why this file (CommonLibrary.psm1) is in the native directory when everything looks to be be pretty common for downloading/extracting files.
2. What the impact of moving functions out of this file and into, for example, utils.ps1
3. Why we went with a psm1 when everything else is in a ps1.

My thought is basically, any time we are trying to get a native tool (I would consider vswhere.exe a native tool), that functionality would be in a script in the native directory, hence why this function is where it is. But I don't have the context on that.

adiaaida commented:
@tmat, I have reverted the change to use get-file, and added a try/catch block for telemetry categorization so they don't show up as unknown failures.

I have opened #4822 to track refactoring common functionality out of common/native.

tmat commented:
@adiaaida Thanks!
